### ISSUE TITLE
repetition_penalty detuning and degrading the audio

### ISSUE THREAD

**Issue by PolyVector** [multimodal-art-projection/YuE#45]
First, thank you for this creating wonderful model, I've been waiting for something like this since Suno/Udio launched last year. It's just amazing!

While tinkering with it, I've noticed repetition_penalty tends to put instruments out of tune and reduce the overall clarity. Repeating chords or bass lines will drift up and down, presumably to avoid repeating tokens too much. Setting repetition_penalty=1 fixes this for me

If I understand YuE architecture correctly, it's generating interleaved tokens to represent the 2 channels: [inst][vocal][inst][vocal][inst][vocal]
The two channels are speaking completely different "languages" as they use different codecs, but repetition_penalty has no concept of this, and treats them all as a single "sentence". That would be another reason to avoid using it.

Anyways, I could be way off base here, just thought I'd share this little observation.

**Comment by a43992899** [comment]
Thanks for the reminder. We adjusted the repetition penalty in the early 0.5B checkpoints because they were producing repetitive outputs. I’ll try your suggestion soon and see if it makes sense to turn it off.

**Comment by hf-lin** [comment]
We have found that setting repetition_penalty to 1.0 makes the generation results better. We have updated the script to make this the default setting. Thank you!

**Comment by PolyVector** [comment]
Sorry to revive this, but I've been noticing that the generations are less creative without the _repetition_penalty_, which I suppose makes sense.  Do you think this might be remedied by increasing the _temperature_, _top_p_ or maybe decreasing the _guidance_?

I'll keep experimenting, this model is occupying all my time, it's just so good!

**Comment by hf-lin** [comment]
How many samples did you evaluate on?

**Comment by PolyVector** [comment]
I've only recently been able to generate longer samples between 1:00 and 1:30, so I've been able to complete maybe 5-10 tracks at that length. I'm randomizing the prompts, so it's very unscientific, just a feeling I had.

**Comment by a43992899** [comment]
We did a 20 sample AB test:

rp=1.0 is generally more stable, but the backing track sounds repetitive sometimes.
rp=1.2 will result in less repetitive backing track pattern, but some silence bad cases.

We will change default rp to 1.1, and add --rp to args.

**Comment by PolyVector** [comment]
Oh you've really tested this, that would melt my computer! Thanks for your time! 🍻

**Comment by PolyVector** [comment]
I've been trying to tune the parameters to my taste and have had incredible luck with this setup:

- **top_k** = 4
- **temperature** = ~1.25
- **top_p** = 0.88 - 0.9
- **repetition_penalty** = 1

Apparently **top_k**+**top_p** is magic 🔥 sauce, at least for me... I suspect that normally the model is allowed too many bad choices and eliminating those helps stabilize the generation. The high **temperature** increases diversity without using **repetition_penalty**, although a low **repetition_penalty** (1.1 or lower) is probably fine.

I was unable to get reliably good generations with only **top_p** or **top_k**, the combination seems to do something special.

**Comment by a43992899** [comment]
Interesting. Thank you for spending time on tuning this and tell us the suggest setup. Let me test it~ @hf-lin

**Comment by PolyVector** [comment]
Oh cool, I hope it's reproducible since my inference setup is completely hacked apart at this point.

I'm not sure that **top_k** is entirely necessary, maybe if **top_p** was perfectly tuned it would accomplish the same thing..?

Feel free to ignore this, but in case anyone is interested, I threw together a LogitsProcessor for repetition penalty that's YuE-friendly. It only looks at the tokens on the same track, so instruments don't affect vocals. I haven't had time to actually test it yet. I assume that as it looks back across song sections they won't be guaranteed to stay on the same even/odd track, so it would require a padding token for any lyric section with an odd number of tokens.

```
# UNTESTED: YuE-friendly penalty for the current channel only...
class OddRepetitionPenaltyLogitsProcessor(LogitsProcessor):
    def __init__(self, penalty: float):
        if not isinstance(penalty, float) or not (penalty > 0):
            raise ValueError(f"`penalty` has to be a strictly positive float, but is {penalty}")

        self.penalty = penalty

    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
        # Create a mask to select only odd tokens in the history
        odd_token_mask = torch.arange(input_ids.shape[1], device=input_ids.device) % 2 == 1

        # Select only odd tokens in the history
        odd_input_ids = input_ids[:, odd_token_mask]

        # Apply the repetition penalty
        score = torch.gather(scores, 1, odd_input_ids)

        # if score < 0 then repetition penalty has to be multiplied to reduce the token probabilities
        score = torch.where(score < 0, score * self.penalty, score / self.penalty)

        # Create a tensor to store the processed scores
        scores_processed = scores.clone()

        # Update the scores of the odd tokens in the history
        scores_processed.scatter_(1, odd_input_ids, score)

        return scores_processed
```

**Comment by WrongProtocol** [comment]
> I've been trying to tune the parameters to my taste and have had incredible luck with this setup:
> 
> * **top_k** = 4
> * **temperature** = ~1.25
> * **top_p** = ~0.9
> * **repetition_penalty** = 1
> 
> Apparently **top_k**+**top_p** is magic sauce, at least for me... I suspect that reducing the model's choices to a few good ones helps stabilize the generation. The high **temperature** increases diversity without using **repetition_penalty**.

I'm running batches of 40 renders at a time ~45 second music each, and these settings made a Huge different in making the outcome more structured and predictable (i.e. usable).
